kernel/scheduler_epoch/src/lib.rs

```diff
@@ -28,92 +36,86 @@ const INITIAL_TOKENS: usize = 10;
 /// An instance of an epoch scheduler, typically one per CPU.
 pub struct Scheduler {
     idle_task: TaskRef,
-    queue: VecDeque<EpochTaskRef>,
+    have_tokens: VecDeque<EpochTaskRef>,
+    out_of_tokens: Vec<EpochTaskRef>,
 }
 
 impl Scheduler {
     /// Creates a new epoch scheduler instance with the given idle task.
     pub const fn new(idle_task: TaskRef) -> Self {
         Self {
             idle_task,
-            queue: VecDeque::new(),
+            have_tokens: VecDeque::new(),
+            out_of_tokens: Vec::new(),
         }
     }
 
-    /// Moves the `TaskRef` at the given `index` in this scheduler's runqueue
-    /// to the end (back) of the runqueue.
-    ///
-    /// Sets the number of tokens for that task to the given `tokens`
-    /// and increments that task's number of context switches.
-    ///
-    /// Returns a cloned reference to the `TaskRef` at the given `index`.
-    fn update_and_move_to_end(&mut self, index: usize, tokens: usize) -> Option<TaskRef> {
-        if let Some(mut priority_task_ref) = self.queue.remove(index) {
-            priority_task_ref.tokens_remaining = tokens;
-            let task_ref = priority_task_ref.task.clone();
-            self.queue.push_back(priority_task_ref);
-            Some(task_ref)
-        } else {
-            None
+    fn try_next(&mut self) -> Option<TaskRef> {
+        while let Some(task) = self.have_tokens.pop_front() {
+            if task.task.is_runnable() {
+                if let Some(task) = self.add_epoch_task(task) {
+                    return Some(task);
+                }
+            } else {
+                task.task
+                    .expose_is_on_run_queue()
+                    .store(false, Ordering::Release);
+                // This check prevents an interleaving where `TaskRef::unblock` wouldn't add
+                // the task back onto the run queue. `TaskRef::unblock` sets the run state and
+                // then checks `is_on_run_queue` so we have to do the inverse.
+                if unlikely(task.task.is_runnable()) {
+                    if let Some(task) = self.add_epoch_task(task) {
+                        return Some(task);
+                    }
+                }
```

Comment:
So, IIUC, the crux of the issue surrounding this `is_on_run_queue()` atomic boolean is that we need to do two things atomically:
1. change the task's runstate to `Runnable` 
2. add the task back to a runqueue

Since we cannot do that, you're using a separate atomic bool to indicate whether a task is on a runqueue. I understand this, but I don't love this design. It's complex without bringing any other benefits, and kind of a messy design choice, as it makes both current and future scheduler policies harder to write.

Ideally this shouldn't be a necessary step. Can we avoid this complexity via an alternative design that's a bit cleaner and doesn't burden the `Scheduler` policy implementor with additional concerns?

Side note: if we actually wanted to store some piece of state inside a task about whether that task is on a runqueue, we might as well store a reference to (or the ID of) the actual runqueue that it's currently on. This would at least make it faster to add the task back once it gets unblocked, which addresses my other comment as a plus.